src/api/TKRecoveryManager.m

```diff
@@ -0,0 +1,189 @@
+    TKRpcErrorHandler *errorHandler;
+    
+    NSString *verificationId;
+    NSString *code;
+    NSString *memberId;
+    TKCrypto *crypto;
+    Alias *alias;
+    MemberRecoveryOperation *recoveryOperation;
+}
+
+- (id)initWithGateway:(GatewayService *)gateway_
+            timeoutMs:(int)timeoutMs_
+         developerKey:(NSString *)developerKey_
+         errorHandler:(TKRpcErrorHandler *)errorHandler_
+               crypto:(id<TKCryptoEngineFactory>)cryptoEngineFactory_ {
+    self = [super init];
+    if (self) {
+        gateway = gateway_;
+        timeoutMs = timeoutMs_;
+        developerKey = developerKey_;
+        errorHandler = errorHandler_;
+        unauthenticatedClient = [[TKUnauthenticatedClient alloc]
+                                 initWithGateway:gateway
+                                 timeoutMs:timeoutMs
+                                 developerKey:developerKey
+                                 errorHandler:errorHandler];
+        cryptoEngineFactory = cryptoEngineFactory_;
+        
+        verificationId = nil;
+        code = nil;
+        memberId = nil;
+        crypto = nil;
+        alias = nil;
+        recoveryOperation = nil;
+    }
+    return self;
+}
+
+- (void)beginRecovery:(NSString *)aliasValue
+            onSuccess:(OnSuccess)onSuccess
+              onError:(OnError)onError {
+    verificationId = nil;
+    code = nil;
+    memberId = nil;
+    crypto = nil;
+    recoveryOperation = nil;
+    alias = [Alias message];
+    alias.value = aliasValue;
+    alias.type = Alias_Type_Unknown;
+    
+    [unauthenticatedClient
+     getTokenMember:alias
+     onSuccess:^(TokenMember *tokenMember) {
+         alias = tokenMember.alias;
+         memberId = tokenMember.id_p;
+         
+         [unauthenticatedClient
+          beginRecovery:alias
+          onSuccess:^(NSString *verificationId_) {
+              verificationId = verificationId_;
+              
+              //Generate keys for later recovery process
+              crypto = [[TKCrypto alloc] initWithEngine:[cryptoEngineFactory createEngine:memberId]];
+              [crypto generateKeys];
+              
+              onSuccess();
+          }
+          onError:onError];
+     }
+     onError:onError];
+}
+
+- (void)verifyRecoveryCode:(NSString *)code
+                 onSuccess:(OnSuccessWithBoolean)onSuccess
+                   onError:(OnError)onError {
+    
+    if (crypto == nil) {
+        onError([NSError
+                 errorFromErrorCode:kTKErrorInvalidRecoveryProcess
+                 details:@"Please call beginRecovery before verifyRecoveryCode"]
+                );
+    }
+    
+    Key *key = [crypto getKeyInfo:Key_Level_Privileged
+                           reason:TKLocalizedString(
+                                    @"Signature_Reason_RecoverMember",
+                                    @"Approve to recover a Token member account")
+                          onError:onError];
+    if (!key) {
+        return;
```

Comment:
error?

Reply:
Error have been propagated by getKeyInfo method whenever key is nil